Provided CA via kumactl

jakubdyszkiewicz opened the pull request:
### Summary

Provided CA via kumactl

Since there is no config for port, it is hardcoded for now.

yskopets commented on app/kumactl/cmd/manage/ca/provided.go:
```diff
@@ -0,0 +1,176 @@
+
+func printListCertificates(certs []types.SigningCert, out io.Writer) error {
+	data := printers.Table{
+		Headers: []string{"ID", "CERT"},
```
It's not helpful for a user to see raw CA cert bytes in the `CERT` column.

A useful info could include:
- Common Name
- Serial Number
- Not valid before
- Not valid after
- SHA-1 fingerprint
- SHA-256 fingerprint

yskopets commented on app/kumactl/cmd/manage/ca/provided_test.go:
```diff
@@ -0,0 +1,172 @@
+		// and
+		Expect(client.addMesh).To(Equal("demo"))
+		Expect(client.addPair).To(Equal(keyPar))
+	})
```
Could we also validate the user feedback message ?

yskopets commented on app/kumactl/cmd/manage/ca/provided_test.go:
```diff
@@ -0,0 +1,172 @@
+		// and
+		Expect(client.deleteCertId).To(Equal("1234-5678"))
+		Expect(client.deleteCertMesh).To(Equal("demo"))
+	})
```
Could we also validate the user feedback message ?

yskopets commented on app/kumactl/cmd/manage/ca/provided_test.go:
```diff
@@ -0,0 +1,172 @@
+
+		// and
+		Expect(client.deleteCaMesh).To(Equal("demo"))
+	})
```
Could we also validate the user feedback message ?

yskopets commented on app/kumactl/cmd/manage/ca/provided.go:
```diff
@@ -0,0 +1,176 @@
+
+func newDeleteCaCmd(pctx *kumactl_cmd.RootContext) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "delete",
```
What is the use case for `kumactl manage ca provided delete` ?
Why `kumactl manage ca provided certificates delete` is not enough ?

jakubdyszkiewicz commented:
Introduced admin in catalog and removed hardcode from kumactl.

For now dataplaneToken api is still in catalog, so it's not incompatibile. Removing it would require to change kumactl and kuma-dp to check if dataplane token ws is on. This will come in a separate PR, but it's not required for 0.3.0 release

yskopets commented on app/kumactl/pkg/cmd/root_context.go:
```diff
@@ -130,51 +133,69 @@ func (rc *RootContext) catalog() (catalog.Catalog, error) {
 }
 
 func (rc *RootContext) CurrentDataplaneTokenClient() (tokens.DataplaneTokenClient, error) {
+	// todo(jakubdyszkiewicz) check enable/disable by checking cp config
```
What does it mean ? Is it different from `if !components.Apis.DataplaneToken.Enabled()` below ?

yskopets replied:
Will be resolved in https://github.com/Kong/kuma/issues/486

yskopets commented on pkg/core/ca/provided/rest/client.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"encoding/json"
 	"fmt"
 
+	kumactl_config "github.com/Kong/kuma/pkg/config/app/kumactl/v1alpha1"
```
Let's avoid dependency ` pkg/core => pkg/config/app/kumactl`.

Probably, we should move this client into `app/kumactl`

yskopets commented on app/kumactl/cmd/config/config_control_planes_add.go:
```diff
@@ -72,7 +72,7 @@ func newConfigControlPlanesAddCmd(pctx *kumactl_cmd.RootContext) *cobra.Command
-	cmd.Flags().StringVar(&args.dataplaneTokenClientCert, "dataplane-token-client-cert", "", "Path to certificate of a client that is authorized to use Dataplane Token Server")
-	cmd.Flags().StringVar(&args.dataplaneTokenClientKey, "dataplane-token-client-key", "", "Path to certificate key of a client that is authorized to use Dataplane Token Server")
+	cmd.Flags().StringVar(&args.adminClientCert, "admin-client-cert", "", "Path to certificate of a client that is authorized to use Admin Server")
+	cmd.Flags().StringVar(&args.adminClientKey, "admin-client-key", "", "Path to certificate key of a client that is authorized to use Admin Server")
```
Let's document this breaking change in the `CHANGELOG.md` (like we did for `0.3.0`)

yskopets commented on pkg/core/bootstrap/autoconfig.go:
```diff
@@ -33,7 +33,12 @@ func autoconfigureCatalog(cfg *kuma_cp.Config) {
 			Url: fmt.Sprintf("http://%s:%d", cfg.General.AdvertisedHostname, cfg.BootstrapServer.Port),
 		},
-		DataplaneToken: catalog.DataplaneTokenApiConfig{},
+		Admin: catalog.AdminApiConfig{
```
> Jakub Dyszkiewicz
answering many questions about this dataplaneToken localurl:
previously the behavior was that if token server is disabled then verification is disabled and dataplaneToken.localUrl in catalog is empty and that’s how we could detect that verification is disabled in kuma-dp and kumactl.
now, we cannot turn off the admin server if verification is disabled and serve admin.localUrl empty because even when verification is turned off someone may use provided ca manager anyways.

yskopets commented on pkg/catalog/types.go:
```diff
@@ -36,6 +42,10 @@ func FromConfig(cfg catalog.CatalogConfig) Catalog {
 				LocalUrl:  cfg.DataplaneToken.LocalUrl,
```
> Jakub Dyszkiewicz
answering many questions about this dataplaneToken localurl:
previously the behavior was that if token server is disabled then verification is disabled and dataplaneToken.localUrl in catalog is empty and that’s how we could detect that verification is disabled in kuma-dp and kumactl.
now, we cannot turn off the admin server if verification is disabled and serve admin.localUrl empty because even when verification is turned off someone may use provided ca manager anyways.

yskopets commented on pkg/config/api-server/catalog/config.go:
```diff
@@ -3,7 +3,8 @@ package catalog
 type CatalogConfig struct {
 	Bootstrap      BootstrapApiConfig
-	DataplaneToken DataplaneTokenApiConfig
+	DataplaneToken DataplaneTokenApiConfig // DEPRECATED: remove in next major version of Kuma
```
> Jakub Dyszkiewicz
answering many questions about this dataplaneToken localurl:
previously the behavior was that if token server is disabled then verification is disabled and dataplaneToken.localUrl in catalog is empty and that’s how we could detect that verification is disabled in kuma-dp and kumactl.
now, we cannot turn off the admin server if verification is disabled and serve admin.localUrl empty because even when verification is turned off someone may use provided ca manager anyways.
